crates/leon/src/template.rs

```diff
@@ -0,0 +1,194 @@
+    /// ```
+    ///
+    /// For an even more ergonomic syntax, see the [`leon::template!`] macro.
+    pub const fn new(items: &'s [Item<'s>], default: Option<&'s str>) -> Template<'s> {
+        Template {
+            items: Cow::Borrowed(items),
+            default: match default {
+                Some(default) => Some(Cow::Borrowed(default)),
+                None => None,
+            },
+        }
+    }
+
+    /// Parse a template from a string.
+    ///
+    /// # Syntax
+    ///
+    /// ```plain
+    /// it is better to rule { group }
+    /// one can live {adverb} without power
+    /// ```
+    ///
+    /// A replacement is denoted by `{` and `}`. The contents of the braces, trimmed
+    /// of any whitespace, are the key. Any text outside of braces is left as-is.
+    ///
+    /// To escape a brace, use `\{` or `\}`. To escape a backslash, use `\\`. Keys
+    /// cannot contain escapes.
+    ///
+    /// ```plain
+    /// \{ leon \}
+    /// ```
+    ///
+    /// The above examples, given the values `group = "no one"` and
+    /// `adverb = "honourably"`, would render to:
+    ///
+    /// ```plain
+    /// it is better to rule no one
+    /// one can live honourably without power
+    /// { leon }
+    /// ```
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use leon::Template;
+    /// let template = Template::parse("hello {name}").unwrap();
+    /// ```
+    ///
+    pub fn parse(s: &'s str) -> Result<Self, ParseError> {
+        Self::parse_items(s).map(|items| Template {
+            items: Cow::Owned(items),
+            default: None,
+        })
+    }
+
+    pub fn render_into(
+        &self,
+        writer: &mut dyn Write,
+        values: &dyn Values,
+    ) -> Result<(), RenderError> {
+        for token in self.items.as_ref() {
+            match token {
+                Item::Text(text) => writer.write_all(text.as_bytes())?,
+                Item::Key(key) => {
+                    if let Some(value) = values.get_value(key) {
+                        writer.write_all(value.as_bytes())?;
+                    } else if let Some(default) = &self.default {
+                        writer.write_all(default.as_bytes())?;
+                    } else {
+                        return Err(RenderError::MissingKey(key.to_string()));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn render(&self, values: &dyn Values) -> Result<String, RenderError> {
+        let mut buf = Vec::with_capacity(
+            self.items
+                .iter()
+                .map(|item| match item {
+                    Item::Key(_) => 0,
+                    Item::Text(t) => t.len(),
+                })
+                .sum(),
+        );
+        self.render_into(&mut buf, values)?;
+
+        // UNWRAP: We know that the buffer is valid UTF-8 because we only write strings.
+        Ok(String::from_utf8(buf).unwrap())
+    }
+
+    pub fn has_key(&self, key: &str) -> bool {
```

Comment:
I think we would also need `fn has_keys(&self, keys: &[&str]) -> bool` in binstalk since we check for `format`, `archive-format` and `archive-suffix`.